tests/runtime/invalid/wasm.rs

```diff
@@ -17,51 +17,163 @@ extern "C" {
-        unreachable!();
     }
     fn invalid_char() {
         unsafe {
             roundtrip_char(0xd800);
         }
         unreachable!();
     }
-    fn invalid_bool() {
+    fn invalid_enum() {
         unsafe {
-            roundtrip_bool(2);
+            roundtrip_enum(400);
         }
         unreachable!();
     }
-    fn invalid_enum() {
+
+    fn test_unaligned() {
+        use imports::{Flag32, Flag64, UnalignedRecord};
+        use std::alloc::{self, Layout};
+        use std::mem;
+        use std::ptr;
+
+        struct Unaligned<T: Copy> {
+            alloc: *mut u8,
+            _marker: std::marker::PhantomData<T>,
+        }
+
+        impl<T: Copy> Unaligned<T> {
+            fn layout() -> Layout {
+                Layout::from_size_align(2 * mem::size_of::<T>(), 8).unwrap()
+            }
+
+            fn new(data: T) -> Unaligned<T> {
+                unsafe {
+                    let alloc = alloc::alloc(Self::layout());
+                    assert!(!alloc.is_null());
+                    ptr::write_unaligned(alloc.add(1).cast(), data);
+                    Unaligned {
+                        alloc,
+                        _marker: Default::default(),
+                    }
+                }
+            }
+
+            fn as_abi(&self) -> (*const T, usize) {
+                unsafe { (self.alloc.add(1).cast(), 1) }
+            }
+        }
+
+        impl<T: Copy> Drop for Unaligned<T> {
+            fn drop(&mut self) {
+                unsafe {
+                    alloc::dealloc(self.alloc, Self::layout());
+                }
+            }
+        }
+
         unsafe {
-            roundtrip_enum(400);
+            let u16s = Unaligned::new(1);
+            let u32s = Unaligned::new(2);
+            let u64s = Unaligned::new(3);
+            let flag32s = Unaligned::new(Flag32::B8);
+            let flag64s = Unaligned::new(Flag64::B9);
+            let records = Unaligned::new(UnalignedRecord { a: 10, b: 11 });
+            let f32s = Unaligned::new(100.0);
+            let f64s = Unaligned::new(101.0);
+            let strings = Unaligned::new("foo");
+            let lists = Unaligned::new(&[102][..]);
+
+            unaligned_roundtrip1(
+                u16s.as_abi().0,
+                u16s.as_abi().1,
+                u32s.as_abi().0,
+                u32s.as_abi().1,
+                u64s.as_abi().0,
+                u64s.as_abi().1,
+                flag32s.as_abi().0,
+                flag32s.as_abi().1,
+                flag64s.as_abi().0,
+                flag64s.as_abi().1,
+            );
+            unaligned_roundtrip2(
+                records.as_abi().0,
+                records.as_abi().1,
+                f32s.as_abi().0,
+                f32s.as_abi().1,
+                f64s.as_abi().0,
+                f64s.as_abi().1,
+                strings.as_abi().0,
+                strings.as_abi().1,
+                lists.as_abi().0,
+                lists.as_abi().1,
+            );
```

Comment:
I think technically to fully test this we'd have to split out these two tests into N different tests, one for each type. Otherwise anything which checks alignment for the `u16s` correctly would trap and then not make it to anything else.

Additionally only the Rust host generator actually checks alignment at this time, changes would be needed for JS & Python to implement the current component model semantics which is to check alignment. I want to get to that eventually but it means that the test here is going to have different behavior in wasmtime and in other hosts for now. 

Given all that maybe this test could just be commented out for now with an issue filed to come back and fix it?

Reply:
https://github.com/bytecodealliance/wit-bindgen/issues/370